src/Contacts.ts

```diff
@@ -0,0 +1,203 @@
+import { Envelope } from '@xmtp/proto/ts/dist/types/message_api/v1/message_api.pb'
+import Client from './Client'
+import { privatePreferences } from '@xmtp/proto'
+import { buildUserPrivatePreferencesTopic } from './utils'
+import { PublishParams } from './ApiClient'
+
+export type AllowListPermissionType = 'allow' | 'block' | 'unknown'
+
+export type AllowListEntryType = 'address'
+
+export class AllowListEntry {
+  value: string
+  entryType: AllowListEntryType
+  permissionType: AllowListPermissionType
+
+  constructor(
+    value: string,
+    entryType: AllowListEntryType,
+    permissionType: AllowListPermissionType
+  ) {
+    this.value = value
+    this.entryType = entryType
+    this.permissionType = permissionType
+  }
+
+  get key(): string {
+    return `${this.entryType}-${this.value}`
+  }
+
+  static fromAddress(
+    address: string,
+    permissionType: AllowListPermissionType = 'unknown'
+  ): AllowListEntry {
+    return new AllowListEntry(address, 'address', permissionType)
+  }
+}
+
+export class AllowList {
+  entries: Map<string, AllowListPermissionType>
+  static _identifier: string
+
+  constructor() {
+    this.entries = new Map<string, AllowListPermissionType>()
+  }
+
+  allow(address: string) {
+    const entry = AllowListEntry.fromAddress(address, 'allow')
+    this.entries.set(entry.key, 'allow')
+    return entry
+  }
+
+  block(address: string) {
+    const entry = AllowListEntry.fromAddress(address, 'block')
+    this.entries.set(entry.key, 'block')
+    return entry
+  }
+
+  state(address: string) {
+    const entry = AllowListEntry.fromAddress(address)
+    return this.entries.get(entry.key) ?? 'unknown'
+  }
+
+  static async getIdentifier(client: Client): Promise<string> {
+    if (!this._identifier) {
+      const { identifier } =
+        await client.keystore.getPrivatePreferencesTopicIdentifier()
+      this._identifier = identifier
+    }
+    return this._identifier
+  }
+
+  static async load(client: Client): Promise<AllowList> {
+    const allowList = new AllowList()
+    const identifier = await this.getIdentifier(client)
+    const envelopes = (
+      await client.listEnvelopes(
+        buildUserPrivatePreferencesTopic(identifier),
+        async ({ message }: Envelope) => {
+          if (message) {
+            const result = await client.keystore.selfDecrypt({
```

Comment:
Same as below. Would be a nice optimization to decrypt in batches. Especially as people's allowlists get longer.